OpenSim/Common/Component.h

```diff
@@ -1577,10 +1595,82 @@ template <class T> friend class ComponentMeasure;
         of StateVariables by name. 
         
         NOTE: If the component name or the state variable name is ambiguous, the 
-         first instance found is returned. To disambiguate use the full name provided
+         an exception is thrown. To disambiguate use the full name provided
          by owning component(s). */
-    const Component* findComponent(const std::string& name, 
-                                   const StateVariable** rsv = nullptr) const;
+    template<class C = Component>
+    const C* findComponent(const std::string& name, 
+                           const StateVariable** rsv = nullptr) const {
+        std::string msg = getConcreteClassName() + "'" + getName() +
+                          "'::findComponent() ";
+        if (name.empty()) {
+            msg += "cannot find a nameless subcomponent.";
```

Comment:
If this is a user-facing message, perhaps something like "ensure all subcomponents have been named" would be friendlier?